### ISSUE TITLE
Allowing window functions to maintain ordering

### ISSUE THREAD

**Issue by roncrdb** [cockroachdb/cockroach#36310]
User is trying to replicate HBase data model on CRDB. 

At the moment, that current implementation of window functions doesn't guarantee preservation of the ordering. 

Having `windower` maintain order could be a possible solution. 

@yuzefovich how could we prioritize this?

**Comment by tim-o** [comment]
Zendesk ticket #[3227](https://cockroachdb.zendesk.com/agent/tickets/3227) has been linked to this issue.

**Comment by plaflamme** [comment]
@yuzefovich I'm happy to help move this forward. I can participate with the required changes or help validate the changes.

**Comment by justinj** [comment]
This is part of #34251.

**Comment by plaflamme** [comment]
@yuzefovich Thanks for the head's up (sorry for the delay GH notifications are failing me at the moment).

I'm obviously not familiar with CDB's distributed SQL engine, but I was hoping that window functions would only need to expose whether or not they're producing stable ordering to the planner. If they are, then the planner could leverage that, and if not it'd have no ability to optimize those steps (the current state of things). This could then be incrementally added to window functions (some would probably be harder than others). I guess I'm assuming that the planner can already optimize out sorting if ordering is guaranteed by certain steps.

FMI: what would be the purpose of maintaining stable ordering at execution time, but not be able to leverage this during planning?

FMI2: is there some documentation or a RFC / specification I can look at for the current implementation plan?

**Comment by yuzefovich** [comment]
@plaflamme yeah, a single GH notification can easily get lost, so no worries there.

1. There would not be any benefit in maintaining stable ordering during execution if the planner cannot use it - both planning and execution have to know about it / implement it for it to work.
2. Unfortunately, there is no documentation. Probably, the closest to that is a draft doc (https://docs.google.com/document/d/1hWfmhIXfIQxAovxZd4NM-6BcwhxQfUDsYevKTiKXDWk/edit?usp=sharing) that I wrote up when distributed window functions PR was being reviewed last Summer. It is, however, very outdated, so I'm not sure whether it's worth reading.

On the bright side, right now there is a lot of progress going on with window functions: the optimizer already supports them (without any optimizations yet though), and I'm working on adding the support to vectorized engine as well.

**Comment by plaflamme** [comment]
@yuzefovich Ok great, sounds like it's moving forward; exciting! Is there an "umbrella" GH issue this work is happening under? I'll go ahead and follow that one.

**Comment by yuzefovich** [comment]
@plaflamme there is no issue that tracks progress in regards to maintaining the stable ordering (apart from this one), but you probably are interested in #34251 and in #37040.
